server/src/main/java/org/elasticsearch/cluster/service/MasterService.java

```diff
@@ -1600,9 +1605,19 @@ public void run(ActionListener<Void> listener) {
                 for (int i = 0; i < entryCount; i++) {
                     final var entry = queue.poll();
                     assert entry != null;
-                    if (entry.acquireForExecution()) {
+                    final var task = entry.taskHolder().getAndSet(null);
```

Comment:
I'm a bit confused as to why in this new logic, we do not call `acquireForExecution()` on the entry, but create a new entry with a new atomic reference?

Reply:
Hmm you're quite right, we should call `entry.acquireForExecution()` here. Fixed in fbc35deb6d4.

We create a new `Entry` because we need to retrieve the task in the ExecutionResult. However I see a problem in this area, since the executing tasks will now no longer return `false` from their `isPending` method. Stand by, I'll test and fix that.